Correctly truncate and deduplicate strings

SR4ven opened the pull request:
Resolves #490

The string primitive will now truncate the test case data in a somewhat intelligent way, so it does no longer yield any duplicates.

As we don't have a list with all values which we can remove duplicates from, only the last value is compared with the current one and is dropped if equal. This requires the fuzz_library to be sorted.

I had to make some changes to the way num_mutation works. It no longer statically calculates all possible mutations as that would have been impossible/way to complex.
Now it generates the "static number of mutations" once and saves it while the "dynamic number of mutations", which depend on the length of the default value, is generated for each call.
However if anyone can thing of a better way to statically generate the correct number of mutations I'm happy to hear.

While writing the tests for the changes, I noticed that `helpers.str_to_bytes()` is incredibly slow. ~23000 took around 20 seconds to convert. Using six as a wrapper for str.encode() I was able to reduce that time down to something like 30ms. It might not have such a big impact when fuzzing a slow target but it's something right?
I chose utf-8 encoding as ascii felt too limiting for the testcases we have. Most characters would render as the default replacement `?`. However, the str_to_bytes function now takes an `encoding` argument which can be used to override the default.

BTW have I ever mentioned that I hate Python2? Just spend over 2 hours to backport this change...

jtpereyda commented:
> As we don't have a list with all values which we can remove duplicates from, only the last value is compared with the current one and is dropped if equal. This requires the fuzz_library to be sorted.

Clever! Nice trick to get a lot of de-dup at little cost.

> I had to make some changes to the way num_mutation works... Now it generates the "static number of mutations" once and saves it while the "dynamic number of mutations", which depend on the length of the default value, is generated for each call.

That'll be an efficiency blip to watch out for in the future if num_mutations gets called frequently. I don't know of a better way, and I don't think num_mutations gets called too many times, not sure though.

> While writing the tests for the changes, I noticed that helpers.str_to_bytes() is incredibly slow. ~23000 took around 20 seconds to convert. Using six as a wrapper for str.encode() I was able to reduce that time down to something like 30ms.

Nice find! Didn't realize str_to_bytes was basically a lousy re-implementation of something six already has.

> BTW have I ever mentioned that I hate Python2? Just spend over 2 hours to backport this change...

😢

jtpereyda commented:
Looks great overall! A few requests and discussion items in the comments. General observation: This is about the most complex block we have (except maybe the recursive block stuff).

SR4ven commented:
>That'll be an efficiency blip to watch out for in the future if num_mutations gets called frequently. I don't know of a better way, and I don't think num_mutations gets called too many times, not sure though.

If I remember correctly it gets called once every test case. So I'd say that is rather frequent :D

Thank you very much for the review! I'll take a closer look at the comments and see if we can optimize anything later.

SR4ven commented on boofuzz/primitives/string.py:
```diff
@@ -223,24 +222,43 @@ def _yield_long_strings(self, sequences):
+            value = default_value * length
+            if value not in self._fuzz_library:
+                yield value
+                if self.max_len is not None and self.max_len <= len(value):
```
The purpose of the max_len checks is to prevent the generation of data which is too long and would be dropped later.
I found the idea of fist generating all possible data and later filtering it really inefficient. Considering that we generate up to 1MB large strings I think this could also have a noticeable performance impact.

But I agree that in some cases it's kinda hard to wrap your head around all these checks and conditions.

So in conclusion, the max_len checks improve the performance at the cost of readability. Not sure if we should remove them.

SR4ven replied:
The last_val duplicate check is intended to filter duplicates in the fuzz_library which are a result of truncation.
It's main purpose is not to filter the data which we generate by multiplying sequences and sizes

SR4ven replied:
In `_yield_long_strings` I didn't want to yield data that would be truncated later. That's why I placed those max_len checks in there and added a break to the for loops to stop unnecessary iterations.

jtpereyda replied:
Works for me. Just to clarify, does this mean that for `_yield_long_strings()`, you won't get a long string exactly equal to the max_len?

jtpereyda replied:
@SR4ven Nevermind, I see now you changed that elsewhere. Good to go!

SR4ven replied:
It will yield all lengths specified in `_long_string_lengths` and `_extra_long_string_lengths` plus one string exactly max_len long from here:
https://github.com/jtpereyda/boofuzz/blob/c338bf69a500018658d528db3432ecd36d4a1217/boofuzz/primitives/string.py#L240-L242

SR4ven commented on boofuzz/primitives/string.py:
```diff
@@ -223,24 +222,43 @@ def _yield_long_strings(self, sequences):
+                self.max_len < min(self._long_string_lengths[0], self._extra_long_string_lengths[0])
+                or self.max_len
+                > max(self._long_string_lengths[-1] + self._long_string_deltas[-1], self._extra_long_string_lengths[-1])
+            ):
```
You are right, that condition doesn't yield a fuzz case if max_len is in the middle of our defined values. Nice find!

Should be fixed in the upcoming commit.

SR4ven commented on boofuzz/primitives/string.py:
```diff
@@ -221,26 +220,49 @@ def _yield_long_strings(self, sequences):
+            if self.max_len is None or size <= self.max_len + 1:
+                s = "D" * size
+                for loc in self.random_indices[size]:
+                    s = s[:loc] + "\x00" + s[loc:]  # Do we really have to insert a char or can we replace it?
```
Do we really have to insert a char or can we replace it at the given location?

jtpereyda replied:
Replacement should work! Probably a silly carry-over as far as I can tell.

SR4ven replied:
Perfect, that will make things a little easier.
I'll change it.

jtpereyda reviewed (commented):
One further Q and one response

SR4ven commented:
Two more small changes:
1. The default encoding for `String()` is now utf-8. This will result in less `?` replacements for non ascii characters. It is also the default encoding in Python 3
2. The string terminator `\x00` no replaces characters at random positions rather than being inserted. All possible positions are now replaced in one test case rather than inserting one by one and yielding each string.

jtpereyda commented:
Sounds good but
> All possible positions are now replaced in one test case rather than inserting one by one and yielding each string.
seems like a more significant change. Instead of the test replacing a single byte several different times, it now replaces a handful of random bytes all at once. The one-byte-at-a-time approach could catch quite a few different failure cases that the several-bytes-at-a-time approach could miss.

SR4ven commented:
The thing is that before we didn’t replace a single byte per case either. We always replaced one byte more than the last case because we overwrite and save the previous string. So with each test case we manipulated one more byte than in the previous one.

But I agree, changing just one byte per case would increase coverage. I’ll try to implement that.

jtpereyda commented:
Hmm that's a great insight, I wonder if that was a bug or the originally intended behavior. Oh well. (If it takes more work than expected we can keep as is and punt)

SR4ven commented:
I noticed that the random_indices generation wasn't very smart. We could end up with duplicates like [1,6,8,6,3]. We just didn't notice because we inserted the terminator at the given index instead of replacing one character.
"01234" - > "01\x00234" - > "01\x00\x00234"
Those cases are of course no duplicates but also not quite what we expected. At least I didn't.

Another thing I noticed is that following case was possible (now with correct replacement):
"DDDDDDDD" -> "D\x00DDDDDD"
"DDDDDDDDDDDDDDDD" -> "D\x00DDDDDDDDDDDDDD"
For every length all possible indices could be chosen for replacement, even if that index was already chosen for a shorter length.
I fixed that by only allowing indices in "new" positions. So in a 16 characters long string only the last 8 characters can be replaced as the first 8 are covered by the 8 character case.

Also, indices are now replaced one at the time to achieve a higher coverage.

jtpereyda commented:
Excellent! Looks good; I think we're ready to merge & release.

SR4ven commented:
Nice, thank you for the review @jtpereyda 

I'll publish the release later on. I'll go for v0.3.0 because of the major mutation iterator changes if you agree.

jtpereyda commented:
Agreed.
